# ãƒ—ãƒ­ã‚°ãƒ©ãƒãƒ¼ï¼ˆProgrammerï¼‰
extends Node


var Department = load("res://ğŸ‹scripts/ğŸª‘grayscale_ğŸ‰visual_novel/department.gd")


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãƒ¡ãƒ¢ãƒªé–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# çŠ¶æ…‹ã€‚ WaitForKeyConfig, KeyConfig, Ready, Main ã®ï¼”ã¤
var current_state = &"WaitForKeyConfig"
# ç¾åœ¨ã®éƒ¨é–€ï¼ˆStringNameå‹ï¼‰
var current_department_name = null
# ç¾åœ¨é³´ã£ã¦ã„ã‚‹èƒŒæ™¯éŸ³æ¥½ã®ãƒãƒ¼ãƒ‰å
var current_bg_music_name = null
# ç¾åœ¨é³´ã£ã¦ã„ã‚‹åŠ¹æœéŸ³ã®ãƒãƒ¼ãƒ‰å
var current_se_name = null

# å…¨è§’æ•°å­—
var zenkaku_numbers = ["ï¼", "ï¼‘", "ï¼’", "ï¼“", "ï¼”", "ï¼•", "ï¼–", "ï¼—", "ï¼˜", "ï¼™"]
# ãƒ‡ãƒ‘ãƒ¼ãƒˆãƒ¡ãƒ³ãƒˆå¤‰æ•°è¾æ›¸ï¼ˆã‚­ãƒ¼ï¼šStringNameå‹ï¼‰
var departments = {}
# ãƒˆæ›¸ãï¼ˆã‚·ãƒŠãƒªã‚ªã®å‘½ä»¤ãƒ‘ãƒ©ã‚°ãƒ©ãƒ•ï¼‰ã§ä½¿ãˆã‚‹å¤‰æ•°ã®è¾æ›¸
var stage_directions_variables = {}
# ç–‘ä¼¼çš„ãªã‚¹ãƒªãƒ¼ãƒ—ã«ä½¿ã†ã‚«ã‚¦ãƒ³ãƒˆ
var sleep_seconds = 0.0


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ã‚¢ãƒ‰ã‚ªãƒ³ãƒ»ã‚¹ãƒ­ãƒƒãƒˆ
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# BGMãƒãƒ¼ãƒ‰ã®ã‚­ãƒ£ãƒƒã‚·ãƒ¥
var bg_musics = null

# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãƒ»ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ãƒ»ãƒ—ãƒ­ã‚°ãƒ©ãƒ ã®ã‚­ãƒ£ãƒƒã‚·ãƒ¥
var message_window_programs = null

# ã‚¤ãƒ©ã‚¹ãƒˆãƒ»ãƒãƒ¼ãƒ‰ã®ã‚­ãƒ£ãƒƒã‚·ãƒ¥
var images = null

# ã‚µã‚¦ãƒ³ãƒ‰ãƒ»ã‚¨ãƒ•ã‚§ã‚¯ãƒˆï¼ˆSound Effect, Sound FXï¼›åŠ¹æœéŸ³ï¼‰ã®ã‚­ãƒ£ãƒƒã‚·ãƒ¥
var sound_fx = null

# ãƒ†ãƒ­ãƒƒãƒ—ã®ã‚­ãƒ£ãƒƒã‚·ãƒ¥
var telops = null


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãƒãƒ¼ãƒ‰ãƒ»ãƒ‘ã‚¹é–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# ãƒ‡ã‚£ãƒ¬ã‚¯ã‚¿ãƒ¼ãƒ»ãƒãƒ–å–å¾—
func monkey():
	return $"ğŸµMonkey"


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# èµ·å‹•æ™‚è¨­å®š
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

func _ready():

	
	self.bg_musics = MonkeyHand.create(
			self.monkey().of_staff().musician_bg_musics())		# æ¢ã™å ´æ‰€

	self.images = MonkeyHand.create(
			self.monkey().of_staff().illustrator())			# æ¢ã™å ´æ‰€

	self.sound_fx = MonkeyHand.create(
			self.monkey().of_staff().musician_sound_fx())			# æ¢ã™å ´æ‰€

	self.telops = MonkeyHand.create(
			self.monkey().of_staff().telop_coordinator())	# æ¢ã™å ´æ‰€


	self.message_window_programs = MonkeyHand.create(
		self.monkey().of_staff().gui_programmer_message_windows())			# æ¢ã™å ´æ‰€


	# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãƒ»ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã«å¯¾å¿œé–¢æ•°ç´ã¥ã‘
	MonkeyHelper.search_node_name_begins_with(
			# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãƒ»ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã®åå‰ã¯ `â– ` ã§å§‹ã¾ã‚‹ã‚‚ã®ã¨ã™ã‚‹
			&"â– ",
			# æ¢ã™å ´æ‰€
			self.monkey().of_staff().gui_programmer_message_windows(),
			func(child_node):
				# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãƒ»ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã®ãƒšãƒ¼ã‚¸é€ã‚Šæ™‚ã€ãƒ‘ãƒ¼ã‚µãƒ¼ã®ãƒ­ãƒƒã‚¯ã‚’è§£é™¤
				child_node.on_message_window_page_forward = func():
					self.monkey().scenario_player().get_current_department_value().set_parse_lock(false))

	# ãƒ‡ãƒ‘ãƒ¼ãƒˆãƒ¡ãƒ³ãƒˆå¤‰æ•°è¾æ›¸ä½œæˆ
	for department_name in self.monkey().get_all_department_names():
		# ç”Ÿæˆ
		var department_value = Department.new()

		# éƒ¨é–€åã‚’ã‚³ãƒ”ãƒ¼
		department_value.name = department_name		# StringName å‹

		# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚’å‡ºåŠ›ã™ã‚‹å¯¾è±¡ã¨ãªã‚‹ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã®åå‰ã€‚ãƒŒãƒ«ã«ã›ãšã€å¿…ãšä½•ã‹å…¥ã‚Œã¦ãŠã„ãŸæ–¹ãŒãƒ‡ãƒãƒƒã‚°ã—ã‚„ã™ã„
		department_value.stack_of_last_displayed_message_window.push_back(&"â– FullScreen")	# StringName å‹ ã‚·ãƒ³ã‚¿ãƒƒã‚¯ã‚¹ãƒ»ã‚·ãƒ¥ã‚¬ãƒ¼

		# å…ˆé ­ã‚»ã‚¯ã‚·ãƒ§ãƒ³ã®åå‰
		department_value.section_name = self.monkey().of_staff().scenario_writer().get_merged_scenario_document(department_name).keys()[0]

		self.departments[department_name] = department_value


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãã®ä»–
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


func _process(delta):

	# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°ãŒå§‹ã¾ã‚‹
	if self.current_state == &"WaitForKeyConfig":
		self.monkey().key_config_node().entry()
		self.current_state = &"KeyConfig"

	# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°ã«åˆ¶å¾¡ã‚’è­²ã‚‹
	elif self.current_state == &"KeyConfig":
		self.monkey().key_config_node().on_process(delta)

	# ä¸»ãªçŠ¶æ…‹ã®å‰ã«
	elif self.current_state == &"Ready":
		self.current_state = &"Main"
		# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
		# æº–å‚™
		# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

		# æœ€åˆã«å®Ÿè¡Œã™ã‚‹éƒ¨é–€å
		self.current_department_name = self.monkey().scenario_writer().department_control().start_department_name

		# ãƒ‘ãƒ¼ã‚¹ã™ã‚‹ãª
		self.monkey().scenario_player().get_current_department_value().set_parse_lock(true)

		# å°æœ¬ã®ã€ŒÂ§ã€ã‚»ã‚¯ã‚·ãƒ§ãƒ³ã®å†ç”Ÿ
		self.monkey().scenario_player().play_section()

		# ä¼è¨€çª“ã‚’ã€ä¸€æ™‚çš„ã«å±…ãªããªã£ã¦ã„ãŸã®ã‚’è§£é™¤ã™ã‚‹
		self.monkey().scenario_player().get_current_message_window_gui().set_appear_subtree(true)

	# ä¸»ãªçŠ¶æ…‹ã«åˆ¶å¾¡ã‚’è­²ã‚‹
	elif self.current_state == &"Main":
		self.monkey().scenario_player().on_process(delta)


# å¤‰æ•°å±•é–‹ã™ã‚‹
# `target_before_change` - ï¼‘è¡Œã‹ã‚‚ã—ã‚Œãªã„ã—ã€æ®µè½ã‹ã‚‚ã—ã‚Œãªã„ã—ã€åŒ¿åé–¢æ•°ã‹ã‚‚ã—ã‚Œãªã„
func expand_variables(target_before_change):
	
	target_before_change = target_before_change.strip_edges()
		
	if target_before_change is String:
		# ã“ã“ã§ `{{banana}} ãªã©ã®å¼•æ•°ã‚’ã€€å¤‰æ•°å±•é–‹ã—ãŸã„
		
		# ç„¡é™ã«å¤‰æ•°å±•é–‹ã™ã‚‹ã“ã¨ã‚’é˜²ããŸã‚ã€å¤‰æ›å‰æ–‡å­—åˆ—ã¨ã€å¤‰æ›å¾Œæ–‡å­—åˆ—ã¯åˆ¥é€”ç”¨æ„ã™ã‚‹
		var terget_after_change = ""
				
		var from = 0
		
		while from < target_before_change.length():
			var open_index = target_before_change.find("{{", from)
			var close_index = target_before_change.find("}}", from)
			
			if from <= open_index and open_index + 2 < close_index:

				# Head
				terget_after_change += target_before_change.substr(from, open_index - from)

				# å¤‰æ•°åå–å¾—
				var key = target_before_change.substr(open_index + 2, close_index - (open_index + 2))
				print("ï¼»ãƒ—ãƒ­ã‚°ãƒ©ãƒãƒ¼ã‚ºãƒ»ãƒãƒ–ã€€å¤‰æ•°å±•é–‹ï¼½ã€€å¤‰æ•°ã‚­ãƒ¼ï¼šï¼»" + key + "ï¼½")
				
				if key in self.monkey().of_staff().programmer().owner_node().stage_directions_variables:
					var value = self.monkey().of_staff().programmer().owner_node().stage_directions_variables[key]
					print("ï¼»ãƒ—ãƒ­ã‚°ãƒ©ãƒãƒ¼ã‚ºãƒ»ãƒãƒ–ã€€å¤‰æ•°å±•é–‹ï¼½ã€€å¤‰æ•°å€¤ï¼šï¼»" + value + "ï¼½")
				
					terget_after_change += value
				
				else:
					print("ï¼»ãƒ—ãƒ­ã‚°ãƒ©ãƒãƒ¼ã‚ºãƒ»ãƒãƒ–ã€€å¤‰æ•°å±•é–‹ï¼½ã€€å¤‰æ•°å€¤ãŒè¦‹ã¤ã‹ã‚Šã¾ã›ã‚“")
					
					# ä»•æ–¹ãªã„ã®ã§ã€ã‚«ãƒƒã‚³æ‚ªã„ãŒã€ã‚­ãƒ¼ï¼ˆãƒãƒ³ãƒ‰ãƒ«ãƒãƒ¼ï¼‰ã‚’ç”»é¢ã«è¡¨ç¤ºã™ã‚‹
					terget_after_change += "{{" + key + "}}"

				from = close_index + 2
		
			else:
				#print("ï¼»åŠ©ç›£ã€€å¤‰æ•°å±•é–‹ï¼½ã€€å¯¾è±¡ãªã—ã€€æ®µè½ï¼šï¼»" + target_before_change + "ï¼½")
				terget_after_change += target_before_change.substr(from)
				break

		return terget_after_change
	
	else:
		#print("ï¼»åŠ©ç›£ã€€å¤‰æ•°å±•é–‹ï¼½ã€€å¯¾è±¡ãªã—ã€€éãƒ†ã‚­ã‚¹ãƒˆ")
		return target_before_change


# ï¼‘ç•ªå¤–å´ã§ãƒ€ãƒ–ãƒ«ã‚¯ã‚©ãƒ¼ãƒ†ãƒ¼ã‚·ãƒ§ãƒ³ãŒæŒŸã‚“ã§ã„ã‚Œã°ã€ãã®ãƒ€ãƒ–ãƒ«ã‚¯ã‚©ãƒ¼ãƒ†ãƒ¼ã‚·ãƒ§ãƒ³ã‚’å¤–ã—ã¾ã™
func trim_double_quotation(line):
	if 2 <= line.length() and line[0]=="\"" and line[-1]=="\"":
		return line.substr(1, line.length()-2)
	
	return line


# å…ˆé ­è¡Œã¨ã€ãã‚Œä»¥å¤–ã«åˆ†ã‘ã¾ã™ã€‚ã§ããªã‘ã‚Œã°ãƒŒãƒ«
func split_head_line_or_tail(text):
	# æœ€åˆã®æ”¹è¡Œã‚’è¦‹ã¤ã‘ã‚‹
	var index = text.find("\n")
	
	if index < 0:
		return null
	
	var head = text.substr(0, index)
	var tail = text.substr(index+1)
	# print("ï¼»åŠ©ç›£ï¼½ã€€headï¼šã€€[" + head + "]")
	# print("ï¼»åŠ©ç›£ï¼½ã€€tailï¼šã€€[" + tail + "]")
	return [head, tail]


# æ•°å€¤ã‚’å…¨è§’ãƒ†ã‚­ã‚¹ãƒˆã«å¤‰æ›
func number_to_zenkaku_text(number, figures):
	var zenkaku_text = ""
	
	while 0 < number:
		var zenkaku_num = self.zenkaku_numbers[number % 10]
		number /= 10
		
		zenkaku_text = zenkaku_num + zenkaku_text
	
	# å·¦å´ã«å…¨è§’ç©ºç™½ã‚’è©°ã‚ã‚‹
	var spaces = figures - zenkaku_text.length()
	for i in range(spaces):
		zenkaku_text = "ã€€" + zenkaku_text
		
	return zenkaku_text


# éƒ¨é–€å¤‰æ•°å–å¾—
func get_department_value(
		department_name):	# StringName
	return self.departments[department_name]
